### ISSUE TITLE
Automatically populate keys in All Tags

### ISSUE THREAD

**Issue by bryceco** [bryceco/GoMap#323]
From https://github.com/bryceco/GoMap/issues/203#issuecomment-621729618:

Personally, I am rarely using the presets tab (sometimes for addresses), but what I would really appreciate: if there would be a button on the "all tags" tab to add an address template (like in presets, fields for street, housenumber, postcode, city and eventually more). 

It would be _nice to have_ a row of buttons in the all tabs view, with custom presets (a button which adds a list of keys and/or values (textfield, or single or list to choose if there are several possibilities) which you have defined yourself, plus eventually a filter which determines when the custom preset button is shown (presence of tags or combinations of tags, e.g. shop=* or amenity in (restaurant, fast_food, pub, cafe) )).

I know there are already custom presets, but they are not accessible from the general tags tab (in autocompletion they are there I think, but not sure), and while it can be filtered when they should be automaticatically displayed, it should allow for combinations (e.g. amenity=<list> and / or shop=*) in the filter, and for setting multiple tags at the same time.

For example I'd like to have a preset that sets "amenity=drinking_water and fountain=nasone" on the same object. Or add "ref:vatin=*" to any object. In the latter case, it would also make sense to have a simple check (number of digits of the value, and whether the first two are capitalized letters). Another combination would be amenity=recycling, defining it also as "container" and "for clothes". I am sure, everybody has some common combinations which they use so often that a personal preset would make sense.

This is the maximum feature description, the "wish you everything" version, of course it could start much smaller ;-) Thanks for your fantastic work!

_Originally posted by @dieterdreist in https://github.com/bryceco/GoMap/issues/203#issuecomment-621729618_

**Comment by bryceco** [comment]
I rarely use All Tags so any kind of feedback is useful. I think the UI for maintaining lots of per-user short cuts might be too onerous for most users, but something that might be a step forward would be to automatically populate the keys that we already have as presets in Common Tags. So a building would automatically have keys (but not values) filled in for all addr:* keys, etc.

**Comment by dieterdreist** [comment]
yes, it's clear that maintaining a lot of individual presets for users is not what most of us will do. On the other hand, many will possbly want to have 1-3 individual presets, and in particular it would be nice to open the custom presets in a way that you could have more than one key set at once.

There is no "one fits all" logic, unfortunately. For example the idea to add address tags to buildings would work mostly ok in Germany (there are either sites or buildings who get the housenumber, depending on the municipality or maybe the "state" (Bundesland)), but it would not work at all in Italy, where every single entrance and gate gets a number assigned hence buildings never get a housenumber tagged (they will typically be accessible through more than just one number). This also breaks inheritance of address information, unfortunately, so we have to repeat addressing on the contained POIs.

**Comment by wtimme** [comment]
As for the filters, in an MVP (for advanced users), we could utilize [Overpass Wizard queries](https://wiki.openstreetmap.org/wiki/Overpass_turbo/Wizard).

The UI would consist of a text field for the query (and a validation of the query) and a list of text-cell-pairs that take the preset‘s tags.

When „All Tags“ is selected for an object, the query is run against the object‘s tags. If they match, a button for applying the preset‘s tags is displayed.

—

I realize this is not something a sporadic mapper would use, sure. But it would give us a version that we could test and reiterate on.

What do you think?

**Comment by bryceco** [comment]
> For example I'd like to have a preset that sets "amenity=drinking_water and fountain=nasone" on the same object.

It seems that an equal solution is to define an iD feature for Nasone that you could select and it would set those tags automatically. 

> Or add "ref:vatin=*" to any object.

This could also be a preset for shops and amenities in EU.

**Comment by dieterdreist** [comment]
the ref:vatin tag is not limited in scope to the EU, although it is mainly used there currently:
https://taginfo.openstreetmap.org/keys/ref:vatin#map

**Comment by bryceco** [comment]
Okay, but I still feel (initially) that it could be a regular preset, especially since you still need to fill in the value yourself.

**Comment by bryceco** [comment]
Please try out the 1.7.8 beta and give feedback here on how All Tags has changed.

**Comment by TheAdventurer64** [comment]
Well, the new tags editor is definitely better, but I have some feedback.

The “Search for Preset” feature could use a overhaul. The preset images look really low quality (Compared to the same photos in the iD editor, especially the phone one) and shouldn’t the Configure button be moved to settings? That would look better.

The “All Choices” section should categorize everything for what they are (One folder for Natural Features, one for shops, and one for amenities.), similar to the old version of Go Map!!

The editor presets should have a feature so that they automatically update whenever a new version of them are released. That way, the developers don’t have to manually import them.

**Comment by bryceco** [comment]
The All Choices comes from iD. I'm not happy about it either but it's the price of sharing the database. 

Changes to the preset database sometimes requires code changes as well, so it's impossible to update it automatically when iD updates it.

**Comment by TheAdventurer64** [comment]
What about the search for preset?

**Comment by github-throwaway** [comment]
@bryceco 

Has there been any progress regarding this? I'm currently logging benches. And would be awesome to have a preset where key and value are already pre-filled for a specific time of bench. It would be so much faster to define it once and then just pick the preset instead of filling every key out again and again. 

That's what I thought the custom preset did anyway. But it seems I was mistaken. What is the custom preset supposed to be used for in its current form?

**Comment by bryceco** [comment]
>  It would be so much faster to define it once and then just pick the preset

This is akin to defining a custom feature, which you could then search for and it would then also show up in the Recently Used features list. I'll try to prototype it.

> What is the custom preset supposed to be used for in its current form?

The custom preset lets you add a new preset field (key/value pair) to selected objects. For example you could add a "name:jp" field to Common Tags if you're often adding japanese names to things.

**Comment by tordans** [comment]
> This is akin to defining a custom feature, which you could then search for and it would then also show up in the Recently Used features list. I'll try to prototype it.

For inspiration: With all those custom things that we add to GoMap, I always wonder if this could be the start of a schema and data that is shared between editors. The OSM user settings (API) would allow to store those things in a way that other editors can access. And those presets, for examples, would be very hand in Every Door, StreetComplete (EE) and iD. It would require, though, that we came up with a shared schema, that all editors use. In the case of the presets, there is a schema to validate against. However I image building a UI for those will be hard. One thing that this shared schema approach would allow is, to move the editing and adding into a external webapp that is separate for the consuming app, making it less of a burden for each of the apps.

Nothing for a prototype, of course, just wanted to mention it for inspiration.

**Comment by bryceco** [comment]
@tordans:

Do other editors support this type of customization currently?

> The OSM user settings (API) would allow to store those things

Probably going to run into the 255 character limit, so need a shared way to store long values. 

> In the case of the presets, there is a schema to validate against. 

There is, but isn’t the schema somewhat complex for people new to it? You have to know what keys are defined by all the various field entries. 

> move the editing and adding into a external webapp

That’s a fantastic idea.

**Comment by bryceco** [comment]
This is added as of 707314770965bd718643261e71e71f127b78e6b6. You define features at the bottom of Common Tags, under 'Custom Features'
